File: 프로그래머스/2/87946. 피로도/피로도.py
# 문제 분석
# 각 던전마다 ['최소 필요 피로도', '소모 피로도']가 있다.
# 하루에 최대한 많은 던전을 방문하고자 한다.

# 접근
# dungeons의 개수는 최대 8이다.
# 순열을 구하면 O(N!) -> O(8!) ~= 40,000
# 가능한 조합을 구하고, 최대 던전 방문 횟수를 계산한다.

candidates = []

def solution(k, dungeons):
    global candidates
    
    answer = -1
    visited = [False] * len(dungeons)
    permutation(dungeons, visited, [])
    
    for cand in candidates:
        tempK = k
        cnt = 0

        for stage in cand:
            if tempK >= stage[0]:
                tempK -= stage[1]
                cnt += 1
                
        answer = max(answer, cnt)

    return answer


def permutation(dungeons, visited, perm):
    global candidates

    if len(perm) == len(dungeons):
        candidates.append(perm[:])
        return
    
    for i in range(len(dungeons)):
        if not visited[i]:
            visited[i] = True
            perm.append(dungeons[i])
            permutation(dungeons, visited, perm)
            perm.pop()
            visited[i] = False
        
        
        
        

File: 프로그래머스/2/87946. 피로도/피로도.swift
// 이해
// 유저의 현재 피로도: k
// 2차원 배열이 주어지고, 배열의 원소는 [최소 필요 피로도, 소모되는 피로도]로 구성
// 원소를 하나 통과할 때마다 "소모되는 피로도"만큼 감소, result += 1
// k < "최소 필요 피로도"인 경우 반복을 종료하고 result를 return
// 순서대로 탐색하는 것이 아닌, 최선의 결과를 만드는 순서로 탐색해야 함.
// 최대 8개 던전의 선택하는 순열의 개수는 8P8 ~= 40,000
// 각 순열을 선형 탐색해야 한다. 순열의 최대 길이는 8이므로, 시간 복잡도는 대략 320,000.

// 의사코드
// let candidateArray = permutation(dungeons, len(dungeons))
// for candidate in candidateArray {}
    // var life = k
    // var tmpResult = 0
    // for element in candidate {}
        // if life >= element[0] {}
            // life -= element[1]
            // tmpResult += 1
        // else { break }
    // result = max(result, tmpResult)
// return result
        
import Foundation

func solution(_ k:Int, _ dungeons:[[Int]]) -> Int {
    
    let candidateArray = permutation(dungeons, dungeons.count)
    var result = 0
    
    for candidate in candidateArray {
        var life = k
        var tmpResult = 0
        for element in candidate {
            if life >= element[0] {
                life -= element[1]
                tmpResult += 1
            } else {
                break
            }
            
        result = max(result, tmpResult)
            
        }
    }
    
    return result
}

func permutation<T>(_ array: [T], _ n: Int) -> [[T]] {
    var result = [[T]]()
    if array.count < n { return result }
    
    var visited = Array(repeating: false, count: array.count)
    
    func cycle(_ now: [T]) {
        if now.count == n {
            result.append(now)
            return
        }
        
        for i in 0..<array.count {
            if visited[i] {
                continue
            } else {
                visited[i] = true
                cycle(now + [array[i]])
                visited[i] = false
            }
        }
    }
    
    cycle([])
    
    return result
}





